src/state.rs

```diff
@@ -201,46 +241,37 @@ impl NetworkState {
         let ret = result.cast().map_err(VMError::from_store_error)?;
 
         // If we reach this point, everything went well and we can use the
-        // updates made in the forked state.
-        *self = fork;
+        // updates made in the forked state. The forked snapshot is set as
+        // the latest, and the previous latest is hashed and put in the map of
+        // snapshots.
+        mem::swap(&mut self.latest, &mut snap);
+
+        let snap_id = Store::hash(&snap.encode_to_vec());
+        self.snapshots
+            .insert(snap_id, snap)
+            .map_err(VMError::from_store_error)?;
 
-        Ok(ret)
+        Ok((ret, snap_id))
     }
 
-    /// Register a host-fn handler
-    pub fn register_host_module<M>(&mut self, module: M)
-    where
-        M: HostModule + 'static,
-    {
-        self.modules
-            .borrow_mut()
-            .insert(module.module_id(), Box::new(module));
+    /// Reverts the latest state to the given `snap_id`.
+    pub fn revert(&mut self, snap_id: &[u8; 32]) -> Result<(), VMError> {
+        let snap = self
+            .snapshots
+            .remove(snap_id)
+            .map_err(VMError::from_store_error)
+            .transpose()
+            .unwrap_or(Err(VMError::UnknownSnapshot))?;
+
```

Comment:
Do we have logic that removes all snapshots more recent than the reverted snapshot? Here we only make the reverted-to snapshot to be the latest, but what about all subsequent snapshots that should be removed?